### PR TITLE
feat: extend errors and their doc around certificate

### PR DESCRIPTION
**Pull request by hadjiszs**
# Description

- add more errors and propagate relevant context with them
- add more rustdoc on errors

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

**Review by iljakuklic — APPROVED** [review]
Looks good. Just some small suggestions.

File: crates/pessimistic-proof/src/proof.rs [hunk 1]
```diff
@@ -5,60 +5,102 @@ use thiserror::Error;
...
-    #[error("Invalid final nullifier root.")]
-    InvalidFinalNullifierRoot,
-    #[error("Invalid imported bridge exit: {0}")]
-    InvalidImportedBridgeExit(#[from] imported_bridge_exit::Error),
-    #[error("Missing token balance proof.")]
-    MissingTokenBalanceProof,
+    /// The previous local exit root declared by the chain does not match the
+    /// one computed by the prover.
+    #[error("Invalid previous local exit root. declared: {declared}, computed: {computed}")]
+    InvalidPreviousLocalExitRoot { declared: Hash, computed: Hash },
```
[the remark this answers is not in the document]
Comment (hadjiszs) [reply crates/pessimistic-proof/src/proof.rs:28]: Do you mean to snapshot test for all of them in the same way as in `agglayer/src/rpc/tests/errors.rs` right?

>if this is or will be exposed through a json API.

Yeah all of them are exposed through the status field of the rpc `getCertificateHeader`

File: crates/agglayer-types/src/lib.rs [hunk 2]
```diff
@@ -42,26 +43,50 @@ pub struct CertificateHeader {
...
+         {computed:?}"
+    )]
     MismatchNewLocalExitRoot { computed: Digest, declared: Digest },
-    #[error("Overflowed imported bridge exits: {0}")]
-    ImportedBridgeExitOverflow(#[from] pessimistic_proof::ProofError),
-    #[error("Failed to apply the Certificate on the given state: {0}")]
-    InvalidCertificateProjection(#[from] pessimistic_proof::utils::smt::SmtError),
+    /// The given token balance got out of bounds.
+    #[error("Token balance got out of bounds.")]
+    BalanceOverflow(#[from] pessimistic_proof::ProofError),
```
[the remark this answers is not in the document]
Comment (hadjiszs) [reply crates/agglayer-types/src/lib.rs]: yeah you're right, I fixed it :+1:

File: crates/pessimistic-proof/src/proof.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (iljakuklic) [reply crates/pessimistic-proof/src/proof.rs:28]: > Do you mean to snapshot test for all of them in the same way as in `agglayer-node/src/rpc/tests/errors.rs` right?

Yeah something like that. Useful for visually inspecting the format.

[the remark this answers is not in the document]
Comment (hadjiszs) [reply crates/pessimistic-proof/src/proof.rs:28]: as discussed in meeting, propagated to https://github.com/agglayer/agglayer/issues/344

**Pull request merged**